jvm-libs/json-rpc/src/test/kotlin/net/consensys/linea/jsonrpc/client/JsonRpcV2ClientImplTest.kt

```diff
@@ -0,0 +1,671 @@
+          assertThat(it.rpcErrorCode).isEqualTo(-32602)
+          assertThat(it.rpcErrorMessage).isEqualTo("Invalid params")
+        }
+      assertThat(retryPredicateCalls[3]).isEqualTo(Ok(value = null))
+      assertThat(retryPredicateCalls[4]).isEqualTo(Ok(value = "SOME RESULT"))
+      assertThat(retryPredicateCalls[5]).isEqualTo(Ok(value = "EXPECTED RESULT"))
+    }
+  }
+
+  @Test
+  fun `when it has connection error propagates to shallRetryRequestPredicate and retries while is true`() {
+    createClientAndSetupWireMockServer(
+      vertx,
+      retryConfig = retryConfig(maxRetries = 10u)
+    ).also { client ->
+      // stop the server to simulate connection error
+      wiremock.stop()
+
+      val retryPredicateCalls = mutableListOf<Result<String?, Throwable>>()
+
+      val reqFuture = client.makeRequest(
+        method = "someMethod",
+        params = emptyList<Any>(),
+        shallRetryRequestPredicate = {
+          retryPredicateCalls.add(it)
+          retryPredicateCalls.size < 2
+        },
+        resultMapper = { it as String? }
+      )
+
+      assertThatThrownBy { reqFuture.get() }
+        .isInstanceOfSatisfying(ExecutionException::class.java) {
+          assertThat(it.cause).isInstanceOfSatisfying(ConnectException::class.java) {
+            assertThat(it.message).contains("Connection refused: localhost/127.0.0.1:")
+          }
+        }
+
+      assertThat(retryPredicateCalls.size).isEqualTo(2)
+      assertThatThrownBy { retryPredicateCalls[0].orElseThrow() }
+        .isInstanceOfSatisfying(ConnectException::class.java) {
+          assertThat(it.message).contains("Connection refused: localhost/127.0.0.1:")
+        }
+    }
+  }
+
+  @Test
+  fun `when it has connection error propagates to shallRetryRequestPredicate and retries until retry config elapses`() {
+    createClientAndSetupWireMockServer(
+      vertx,
+      retryConfig = retryConfig(maxRetries = 2u, timeout = 8.seconds, backoffDelay = 5.milliseconds)
+    ).also { client ->
+      // stop the server to simulate connection error
+      wiremock.stop()
+
+      val retryPredicateCalls = mutableListOf<Result<String?, Throwable>>()
+
+      val reqFuture = client.makeRequest(
+        method = "someMethod",
+        params = emptyList<Any>(),
+        shallRetryRequestPredicate = {
+          retryPredicateCalls.add(it)
+          true // keep retrying
+        },
+        resultMapper = { it as String? }
+      )
+
+      assertThatThrownBy { reqFuture.get() }
+        .isInstanceOfSatisfying(ExecutionException::class.java) {
+          assertThat(it.cause).isInstanceOfSatisfying(ConnectException::class.java) {
+            assertThat(it.message).contains("Connection refused: localhost/127.0.0.1:")
+          }
+        }
+      assertThat(retryPredicateCalls).hasSizeBetween(1, 3)
+    }
+  }
+
+  @Test
+  fun `when shared predicate is defined shall retry when any of then returns true`() {
```

Comment:
```suggestion
  fun `when shared predicate is defined shall retry when any of them returns true`() {
```

Reply:
I don't understand purpose of this test TBH. What is a shared predicate? I can only see 1 usage of it, so to me it's exclusive, not shared

Reply:
"shared" its an **optional predicate** that's is passed to the constructor that is applied to every request before the predicate  passed in `makeRequest`. The intent is to pass generic predicates in the constructor for generic errors as Connection error, etc..